apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+//
+// Implementations can define their own protocols if a core ProtocolType does not
+// exist. Such definitions must use prefixed name, such as
+// `mycompany.com/my-custom-protocol`. Un-prefixed names are reserved for core
+// protocols. Any protocol defined by implementations will fall under custom
+// conformance.
+type ProtocolType string
+
+const (
+	// Accepts cleartext HTTP/1.1 sessions over TCP.
+	HTTPProtocolType ProtocolType = "HTTP"
+
+	// Accepts HTTP/1.1 or HTTP/2 sessions over TLS.
+	HTTPSProtocolType ProtocolType = "HTTPS"
+
+	// Accepts TLS sessions over TCP.
+	TLSProtocolType ProtocolType = "TLS"
+
+	// Accepts TCP sessions.
+	TCPProtocolType ProtocolType = "TCP"
+
+	// Accepts UDP packets.
+	UDPProtocolType ProtocolType = "UDP"
+)
+
+// GatewayTLSConfig describes a TLS configuration.
+type GatewayTLSConfig struct {
+	// Mode defines the TLS behavior for the TLS session initiated by the client.
+	// There are two possible modes:
+	// - Terminate: The TLS session between the downstream client
+	//   and the Gateway is terminated at the Gateway. This mode requires
+	//   certificateRef to be set.
+	// - Passthrough: The TLS session is NOT terminated by the Gateway. This
+	//   implies that the Gateway can't decipher the TLS stream except for
+	//   the ClientHello message of the TLS protocol.
+	//   CertificateRef field is ignored in this mode.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=Terminate
+	Mode *TLSModeType `json:"mode,omitempty"`
+
+	// CertificateRef is a reference to a Kubernetes object that contains a TLS
+	// certificate and private key. This certificate is used to establish a TLS
+	// handshake for requests that match the hostname of the associated
+	// listener.
+	//
+	// References to a resource in different namespace are invalid UNLESS there
+	// is a ReferencePolicy in the target namespace that allows the certificate
+	// to be attached. If a ReferencePolicy does not allow this reference, the
+	// "ResolvedRefs" condition MUST be set to false for this listener with the
+	// "InvalidCertificateRef" reason.
+	//
+	// This field is required when mode is set to "Terminate" (default) and
+	// optional otherwise.
+	//
+	// CertificateRef can reference a standard Kubernetes resource, i.e. Secret,
+	// or an implementation-specific custom resource.
+	//
+	// Support: Core (Kubernetes Secrets)
+	//
+	// Support: Implementation-specific (Other resource types)
+	//
+	// +optional
+	CertificateRef *ObjectReference `json:"certificateRef,omitempty"`
+
+	// Options are a list of key/value pairs to give extended options
+	// to the provider.
+	//
+	// There variation among providers as to how ciphersuites are
+	// expressed. If there is a common subset for expressing ciphers
+	// then it will make sense to loft that as a core API
+	// construct.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	Options map[string]string `json:"options,omitempty"`
```

Comment:
I understand the need for this. But we need to be very clear on how this field is validated. max key count. max key length. max value length.

Reply:
Added max key count in #772, I think the only way to accomplish max key or value length is to use a custom type (which we probably need to do).